### ISSUE TITLE
AttributeCollection[Type type] uses a static lock

### ISSUE THREAD

**Issue by seanhalliday** [dotnet/runtime#99178]

**Comment by huoyaoyuan** [comment]
The only static state to protect is `s_defaultAttributes`, which is already protected in `GetDefaultAttribute`. It should really be a `ConditionalWeakTable` in order to support collectible types.

**Comment by steveharter** [comment]
Code: https://github.com/dotnet/runtime/blob/49a8bb7eb31f2aea9c895f4e9f6c2f6567281bdd/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/AttributeCollection.cs#L27

**Comment by steveharter** [comment]
> The only static state to protect is s_defaultAttributes, which is already protected in GetDefaultAttribute

Yes `s_internalSyncObject` is used in part to protect `s_defaultAttributes` so using a static makes sense.

> It should really be a ConditionalWeakTable in order to support collectible types.

Is the underlying concern\issue around unloadable types?

**Comment by huoyaoyuan** [comment]
> Is the underlying concern\issue around unloadable types?

The `Type` and attribute instance will permanently stay in the `s_defaultAttributes` dictionary. This prevents the type and the assembly to be collected.
